mmaprototype: add more error cases in TestNormalizedVoterAllRelationships

wenyihu6 opened the pull request:
Based on top of https://github.com/cockroachdb/cockroach/pull/158393. Only the last 5 commits. 

---

Epic: CRDB-55052 
Release note: none

---
**mmaprototype: add more error cases test coverage**


---
**mmaprototype: improve comment on relationship**


---
**mma: additional case for non-intersecting conjunctions**

A=[+region=a, +zone=a1], B=[+region=a, +zone=a2] is now classified as
non-intersecting. This fixes an existing todo.

---
**mmaprototype: fix conjNonIntersecting**

Previously, we started returning conjNonIntersecting when we encounter
same type, same key, different values. However, the fix only returned
early if the differing value is encountered from cc. This commit changed
it so that we also check and return early for b.

---
**mmaprototype: add test cases for conjNonIntersecting**

wenyihu6 committed:
mmaprototype: add assertion on emptyVoterConstraintIndex

This commit adds assertions on emptyVoterConstraintIndex and
emptyConstraintIndex since there should only be one.

wenyihu6 committed:
mmaprototype: clean up doStructuralNormalization

This commit refactors some code in doStructuralNormalization to
use min utility and also adds assertions on programming errors.

wenyihu6 committed:
mmaprototype: use >= 0 for emptyVoterConstraintIndex

This commit fixes >= 0 when checking for whether there were
valid emptyVoterConstraintIndex and emptyConstraintIndex
since 0 is a valid index.

wenyihu6 committed:
fixup! mmaprototype: add assertion on emptyVoterConstraintIndex

wenyihu6 committed:
fixup! mmaprototype: use >= 0 for emptyVoterConstraintIndex

wenyihu6 committed:
mmaprototype: extract relationship building between voter and all-replica constraints

sumeerbhola committed:
mma: additional case for non-intersecting conjunctions

A=[+region=a, +zone=a1], B=[+region=a, +zone=a2] is now classified as
non-intersecting. This fixes an existing todo.

Epic: CRDB-55052

Release note: None

# Conflicts:
#	pkg/kv/kvserver/allocator/mmaprototype/constraint.go

wenyihu6 committed:
mmaprototype: address pr comments

wenyihu6 committed:
fixup! mmaprototype: address pr comments

wenyihu6 committed:
mmaprototype: improve comment on relationship

wenyihu6 committed:
mmaprototype: rename TestVoterAndAllRelationshipTable

wenyihu6 committed:
mmaprototype: print normalized in TestNormalizedVoterAllRelationships

wenyihu6 committed:
mmaprototype: rename file to be testname

wenyihu6 committed:
mmaprototype: add comment for makeBasicNormalizedSpanConfig

wenyihu6 committed:
mmaprototype: add more error cases test coverage

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/158656)
<!-- Reviewable:end -->

wenyihu6 committed:
mmaprototype: fix conjNonIntersecting

Previously, we started returning conjNonIntersecting when we encounter
same type, same key, different values. However, the fix only returned
early if the differing value is encountered from cc. This commit changed
it so that we also check and return early for b.

wenyihu6 committed:
mmaprototype: add test cases for conjNonIntersecting

wenyihu6 commented:
Closing in favor of https://github.com/cockroachdb/cockroach/pull/158722. 

Opened a new PR since this one hasn’t been reviewed yet, which should make reviewable cleaner given my rebase.
